src/sql/workbench/services/connection/browser/connectionManagementService.ts

```diff
@@ -874,6 +876,61 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 		return false;
 	}
 
+	/**
+	 * Refresh Azure access token if it's expired.
+	 * @param uri connection uri
+	 * @returns true if no need to refresh or successfully refreshed token
+	 */
+	public async refreshAzureAccountTokenIfNecessary(uri: string): Promise<boolean> {
+		let profile = this._connectionStatusManager.getConnectionProfile(uri);
+		if (!profile) {
+			this._logService.warn(`Connection not found for uri ${uri}`);
+			return false;
+		}
+
+		//wait for the pending reconnction promise if any
+		let previousReconnectPromise = this._uriToReconnectPromiseMap[uri];
+		if (previousReconnectPromise) {
+			this._logService.info(`Found pending reconnect promise for uri ${uri}, waiting.`);
+			try {
+				let previousConnectionResult = await previousReconnectPromise;
+				if (previousConnectionResult && previousConnectionResult.connected) {
+					this._logService.info(`Previous pending reconnection for uri ${uri} succeeded.`);
+					return true;
+				}
+				this._logService.info(`Previous pending reconnection for uri ${uri} failed with error ${previousConnectionResult.errorMessage}, error code: ${previousConnectionResult.errorCode}.`);
+			} catch (err) {
+				this._logService.info(`Previous pending reconnect promise for uri ${uri} is rejected with error ${err}, will attempt to reconnect if necessary.`);
+			}
+		}
+
+		let expiry = profile.options.expiresOn;
+		if (typeof expiry === 'number' && !Number.isNaN(expiry)) {
+			const currentTime = new Date().getTime() / 1000;
+			const maxTolerance = 2 * 60; // two minutes
+			if (expiry - currentTime < maxTolerance) {
+				this._logService.info(`Access token expired for connection ${profile.id} with uri ${uri}`);
+				try {
+					let connectionResultPromise = this.connect(profile, uri);
+					this._uriToReconnectPromiseMap[uri] = connectionResultPromise;
+					let connectionResult = await connectionResultPromise;
+					delete this._uriToReconnectPromiseMap[uri];
+					if (!connectionResult || !connectionResult.connected) {
+						this._logService.error(`Failed to refresh connection ${profile.id} with uri ${uri}, result: ${connectionResult}`);
+						await this.disconnect(profile);
```

Comment:
What's the reason for disconnecting here? That doesn't seem like it'd actually help anything and would just make the rest of the flow that's trying to refresh the connection probably fail in an odd way. Throwing an exception here being very clear about the issue (failed to refresh) seems like more the approach I think we'd want to take

Reply:
I was leaning towards asking users to start a new connection for whatever scenario they are using if a reconnection attempt failed so I chose to delete the connection and removing the related profile. I can change this to throw an exception if that's preferred.

Reply:
I think that would be better - basically treating it like any other connection error and letting the user attempt to reconnect if they want. Let's just make sure that the message displayed is clear enough about what went wrong.

Reply:
Sure, I just changed the code to throw the error with error msg and error code.